File: typescript/packages/mechanisms/evm/src/batch-settlement/client/fileStorage.ts
import { unlink } from "node:fs/promises";
import { join } from "node:path";

import { isNodeEnoent, readJsonFile, writeJsonAtomic } from "../storage-utils";
import type { FileChannelStorageOptions } from "../types";
import type { ClientChannelStorage, BatchSettlementClientContext } from "./storage";

/**
 * Node.js file-backed {@link ClientChannelStorage} for the batched client scheme.
 * Each channel's context is persisted as `{root}/client/{channelId}.json` so that channel
 * records survive process restarts.
 */
export class FileClientChannelStorage implements ClientChannelStorage {
  private readonly root: string;

  /**
   * Creates file-backed client channel storage under the given root directory.
   *
   * @param options - Configuration including the storage root directory.
   */
  constructor(options: FileChannelStorageOptions) {
    this.root = options.directory;
  }

  /**
   * Loads the stored client context for a channel, if present.
   *
   * @param key - Channel storage key (typically a lowercased channelId).
   * @returns Parsed context or `undefined` when the file is missing.
   */
  async get(key: string): Promise<BatchSettlementClientContext | undefined> {
    return readJsonFile<BatchSettlementClientContext>(this.filePath(key));
  }

  /**
   * Persists the client context for a channel.
   *
   * @param key - Channel storage key.
   * @param context - Context record to write.
   */
  async set(key: string, context: BatchSettlementClientContext): Promise<void> {
    await writeJsonAtomic(this.filePath(key), context);
  }

  /**
   * Removes the persisted context file for a channel, if it exists.
   *
   * @param key - Channel storage key.
   */
  async delete(key: string): Promise<void> {
    try {
      await unlink(this.filePath(key));
    } catch (err: unknown) {
      if (isNodeEnoent(err)) return;
      throw err;
    }
  }

  /**
   * Absolute path to the JSON file for a channel.
   *
   * @param key - Channel storage key.
   * @returns Filesystem path under `{root}/client/...`.
   */
  private filePath(key: string): string {
    return join(this.root, "client", `${key.toLowerCase()}.json`);
  }
}

File: python/x402/mechanisms/evm/batch_settlement/client/file_storage.py
"""File-backed client channel storage for the batch-settlement scheme."""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path

from ..storage_utils import delete_file, read_json_file, write_json_atomic
from .storage import BatchSettlementClientContext, ClientChannelStorage


@dataclass
class FileChannelStorageOptions:
    """Configuration for a file-backed channel storage."""

    directory: str | Path


class FileClientChannelStorage(ClientChannelStorage):
    """File-backed client channel storage.

    Each channel's context is persisted to `{root}/client/{channelId}.json` so
    channel records survive process restarts.
    """

    def __init__(self, options: FileChannelStorageOptions):
        self._root = Path(options.directory)

    def get(self, key: str) -> BatchSettlementClientContext | None:
        data = read_json_file(self._file_path(key))
        if data is None:
            return None
        return BatchSettlementClientContext.from_dict(data)

    def set(self, key: str, context: BatchSettlementClientContext) -> None:
        write_json_atomic(self._file_path(key), context.to_dict())

    def delete(self, key: str) -> None:
        delete_file(self._file_path(key))

    def _file_path(self, key: str) -> Path:
        return self._root / "client" / f"{key.lower()}.json"


__all__ = ["FileChannelStorageOptions", "FileClientChannelStorage"]
